release_util/management/commands/__init__.py

```diff
@@ -132,164 +134,215 @@ def state(self):
-    def __init__(self, stderr, database_name):
-        self.to_apply = []
-        self.migration_state = {
-            'success': [],
-            'failure': None,
-            'unapplied': [],
-            'rollback_commands': [],
-        }
-        self.timer = default_timer
-        self.stderr = stderr
-        self.database_name = database_name
 
-        # Execute command to find/record the unapplied migrations.
-        out = StringIO()
-        call_command("show_unapplied_migrations",
-                     interactive=False,
-                     stdout=out,
-                     database=self.database_name,
-                     standalone=False)
+    If you want to run specific migrations, you can either initialize it with
+    a list of migration tuples, or add migrations by calling add_migrations(),
+    then call apply().
 
-        # Load the passed-in YAML into a dictionary.
-        self.input_migrations = yaml.safe_load(out.getvalue())
+    If you want to run all migrations, simply call apply_all(). Calling
+    apply_all() while a MigrationSession has migrations added to it will raise
+    an error.
 
-        # Build a list of migrations that will be applied.
-        for migration in self.input_migrations['migrations']:
-            self.to_apply.append(migration)
+    Once either apply() or apply_all() has been called, calling apply(),
+    apply_all(), or add_migrations() will raise an error.
+
+    The "stderr" argument is deprecated and not used.
+    """
+    def __init__(self, stderr, database_name, migrations=None):
+        self._to_apply = []
+        self._migration_state = []
+        self._timer = default_timer
+        self._closed = False
+
+        self._database_name = database_name
+        if migrations:
+            self.add_migrations(migrations)
 
         # Regex built to match migration output like this line:
         #    Applying release_util.0001_initial...  OK
         self.migration_regex = re.compile(r'Applying (?P<app_name>[^\.]+)\.(?P<migration_name>[^\.]+)[\. ]+(?P<success>[OK]*)$')
 
-    def more_to_apply(self):
+    def add_migrations(self, migrations):
         """
-        Returns True if more migrations remain to apply in this session, else False.
+        Add migrations to be applied. Migrations are of the form [(app, migration_name), ...]
         """
-        return len(self.to_apply) > 0
+        if self._closed:
+            raise MigrationSessionError("Can't change applied session")
+        self._to_apply.extend(migrations)
 
-    def _add_rollback_commands(self):
+    def _gather_unapplied_migrations(self, loader):
         """
-        Generate rollback commands for the apps that have had migrations applied.
-        If an app's migration has failed, include that rollback command as well.
+        Output a list of unapplied migrations in the form [['migration1', migration2'], ...].
+        This implementation is mostly copied from the Django 'showmigrations' mgmt command.
```

Comment:
I don't actually know what the args are since I just copy-pasted this from `show_unapplied_migrations.py`. @doctoryes?